### ISSUE TITLE
Improve upon the current TTS system

### ISSUE THREAD

**Issue by JaytleBee** [unitystation/unitystation#1525]
# Description
The way TTS currently works is that every client connected to a server with TTS enabled sends a request to the server at unitystation.org to request an audio file of the chat line. This is not only a system with a single point of failure, but it's also not scalable. The more developers and players we get, the worse the situation becomes. This is obviously not ideal.

I talked about this with chairbender in the discord and we came up with the idea of using MaryTTS (which we seem to be using already) locally somehow. Of course, this opens up the can of worms called "Java", but I think that kind of solution is still better than what we have right now. It seems @DooblyNoobly and @fomalsd set this up so please comment here if I'm missing something.

## Category
TODO

**Comment by fomalsd** [comment]
we tried doing something with IKVM.net, but it didn't work out: it's discontinued and javax.sound.* is not supported at all:
https://sourceforge.net/p/ikvm/wiki/Limits/

**Comment by chairbender** [comment]
Yeah, we were thinking that we shouldn't need IKVM. Just run the server as a separate java process using the actual JRE. Will need some logic to handle initial setup / startup, and checking if Java is installed or bundling the JRE with the game, etc...but seems like it should be doable and cross platform.

**Comment by fomalsd** [comment]
Things to keep in mind:
"TTS is not really used much atm and bundling java+marytts would add like +100-150mb while compressed game is 60mb, plus it won't work for android+ios obviously"
Of course I'd **much** rather prefer a local solution, but please consider priorities and convenience, too

**Comment by chairbender** [comment]
seems there isn't a free cross platform TTS, you'd need to use an approach that worked for each supported platform. This guy outlines how it could be done:
https://stackoverflow.com/a/42576071/765419

It seems like a lot of work, but the bulk of the work is really just hooking up the TTS for each platform - most of the work of actually implementing TTS is probably already done for each platform, just have to hunt it down.

And then once it's implemented we get to have the honor of being the first to have an open source, cross platform client-side TTS in Unity!

**Comment by fomalsd** [comment]
meh
I've seen and analyzed a paid asset that does exactly as you described but the dull part is all the voices would be different for each system, that sucks
MaryTTS provides several tweakable voices that sound good, and I really want them to be consistent regardless of OS

**Comment by chairbender** [comment]
I think those are nice to have features, but it sounds like you didn't think marytts is an option because it doesn't work on Android and ios and would increase the game size. So I'm trying to think of an alternative.

**Comment by fomalsd** [comment]
We could try converting MaryTTS jar into a native library using GraalVM and see if it works:
https://github.com/oracle/graal/issues/1786
Earlier we tried doing something similar with IKVM.Net but it lacked javax.sound support, so sound generation didn't work (web server part worked though)

**Comment by fomalsd** [comment]
While GraalVM native image compiler is quite painful to set up (unlike IKVM.Net which didn't require any configs), it might eventually work (for desktop)
Waiting for https://github.com/oracle/graal/issues/2063 to be resolved to proceed further

**Comment by TheBricktop** [comment]
There is also a wrapper for espeak for net , You could try that , 
https://github.com/rhdunn/espeak

https://github.com/unitycoder/UnityRuntimeTextToSpeech/blob/master/README.md

**Comment by MaxIsJoe** [comment]
We've added a new TTS system that's better than the old one and more cost effective. Closing this.
